aten/src/ATen/native/quantized/cpu/ACLUtils.h

```diff
@@ -0,0 +1,251 @@
+#pragma once
+
+#include <ATen/Config.h>
+#if defined(__aarch64__) && AT_MKLDNN_ACL_ENABLED()
+
+#include <ATen/native/quantized/cpu/OnednnUtils.h>
+#include <arm_compute/core/Error.h>
+#include <arm_compute/core/TensorInfo.h>
+#include <arm_compute/function_info/ActivationLayerInfo.h>
+#include <arm_compute/runtime/Allocator.h>
+#include <arm_compute/runtime/NEON/functions/NEActivationLayer.h>
+#include <arm_compute/runtime/NEON/functions/NEGEMMLowpMatrixMultiplyCore.h>
+#include <arm_compute/runtime/NEON/functions/NEQuantizationLayer.h>
+#include <arm_compute/runtime/Tensor.h>
+#include <array>
+
+using ACLDynamicQuantMatmulCacheKey = std::tuple<
+    int64_t, // M
+    bool, // FUSE_RELU
+    int64_t // NUM_THREADS
+    >;
+
+enum ACLDynamicQuantMatmulCacheKeyIndex {
+  M,
+  FUSE_RELU,
+  NUM_THREADS,
+};
+
+struct ACLDynamicQuantMatmul {
+  arm_compute::Tensor src_s8_tensor;
+  arm_compute::Tensor src_fp32_tensor;
+  arm_compute::Tensor wei_tensor;
+  arm_compute::Tensor bia_tensor;
+  arm_compute::Tensor dst_tensor;
+  arm_compute::NEQuantizationLayer quant;
+  std::shared_ptr<arm_compute::IMemoryManager> memory_manager{
+      arm_compute::MemoryManagerOnDemand::make_default()};
+  arm_compute::NEGEMMLowpMatrixMultiplyCore gemm{memory_manager};
+  arm_compute::NEActivationLayer acl_relu;
+  // configuration details for the ACL gemm
+  arm_compute::TensorInfo src_s8_tensor_info;
+  arm_compute::TensorInfo src_fp32_tensor_info;
+  arm_compute::TensorInfo wei_tensor_info;
+  arm_compute::TensorInfo bia_tensor_info;
+  arm_compute::TensorInfo dst_tensor_info;
+  arm_compute::GEMMInfo gemm_info;
+  arm_compute::ActivationLayerInfo acl_relu_info{
+      arm_compute::ActivationFunction::RELU};
+  bool with_bias{false};
+
+  // key for use in the cache
+  ACLDynamicQuantMatmulCacheKey key;
+
+  ~ACLDynamicQuantMatmul() {
+    // this will free memory allocated for the quantized src tensor since the
+    // allocation happened through ACL: src_s8_tensor.allocator()->allocate()
+    src_s8_tensor.allocator()->free();
+    // this will not free memory, it will just tell ACL that we're no longer
+    // using the pointer
+    wei_tensor.allocator()->free();
+    if (with_bias) {
+      bia_tensor.allocator()->free();
+    }
```

Comment:
Wouldn't it be better to express something like that by defining bias tensor as `std::optional<arm_compute::Tensor> bias_tensor;`?

Reply:
Done, I now use `std::optional` for `bia_tensor` amd `bia_tensor_info`